### PR TITLE
Define an interface for the path planner and create a placeholder path planner

### PR DESCRIPTION
**Pull request by jordanzliu**
### Description

Define an interface for the path planner and create a placeholder path planner

### Testing Done

Interface does not require testing. 1 unit test for placeholder path planner and placeholder violation 
function combined.

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/thunderbots/software/ai/navigator/path_planner/straight_line_path_planner.cpp [hunk 1]
```diff
@@ -0,0 +1,12 @@
+//
+// Created by jordan on 3/23/19.
+//
+
+#include "straight_line_path_planner.h"
```
Comment (jonathanlew) [inline src/thunderbots/software/ai/navigator/path_planner/straight_line_path_planner.cpp]: I think this should be ai/navigator/path_planner/straight_line_path_planner.h

File: src/thunderbots/software/ai/navigator/path_planner/path_planner.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+     * @param obstacles a vector of obstacles
+     * @param violation_function a function that returns the distance
+     *                          that a point is violating a boundary by
+     * @return
+     */
+    virtual std::optional<std::vector<Point>> findPath(
+        const Point &start, const Point &dest, const std::vector<Obstacle> &obstacles,
+        const ViolationFunction &violation_function) = 0;
+    virtual ~PathPlanner()                           = default;
+};
```
Comment (garethellis0) [inline src/thunderbots/software/ai/navigator/path_planner/path_planner.h]: github complaining about end-of-file

File: src/thunderbots/software/ai/navigator/path_planner/straight_line_path_planner.cpp [hunk 3]
```diff
@@ -0,0 +1,12 @@
+//
+// Created by jordan on 3/23/19.
+//
```
Comment (garethellis0) [inline src/thunderbots/software/ai/navigator/path_planner/straight_line_path_planner.cpp]: We should switch this to a proper full start-of-file comment

File: src/thunderbots/software/ai/navigator/path_planner/straight_line_path_planner.h [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+     * @param start start point
+     * @param dest destination point
+     * @param obstacles unused parameter
+     * @param violation_function unused parameter
+     * @return a vector that is {start, dest}
+     */
+    std::optional<std::vector<Point>> findPath(
+        const Point &start, const Point &dest, const std::vector<Obstacle> &obstacles,
+        const ViolationFunction &violation_function) override;
+};
```
Comment (garethellis0) [inline src/thunderbots/software/ai/navigator/path_planner/straight_line_path_planner.h]: github complaining about end-of-file

File: src/thunderbots/software/ai/navigator/path_planner/violation_function.h [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    /**
+     * Returns the distance that a point is violating a boundary
+     * by, in metres.
+     * @param point a Point
+     * @return the distance that a point is violating a boundary
+     * by, in metres.
+     */
+    virtual double getViolationForPoint(const Point& point) = 0;
+    virtual ~ViolationFunction()                            = default;
+};
```
Comment (garethellis0) [inline src/thunderbots/software/ai/navigator/path_planner/violation_function.h]: github complaining about end-of-file

File: src/thunderbots/software/ai/navigator/path_planner/violation_function.h [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+#include "geom/point.h"
+
+/**
+ * The ViolationFunction is an object that represents a function
+ * that can be called to evaluate the boundary violation of a
+ * point on the field, or implement other soft avoidance areas
+ * on the field.
+ */
+
+class ViolationFunction
```
Comment (garethellis0) [inline src/thunderbots/software/ai/navigator/path_planner/violation_function.h]: Hmm, I'd almost change this to something like 
```
using ViolationFunction = std::function<double(Point)>;
```
instead? (Not totally sure if my code is right, but you get the idea).

File: src/thunderbots/software/test/ai/navigator/path_planner/straight_line_path_planner.cpp [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    EXPECT_EQ(path_points[0], start);
+    EXPECT_EQ(path_points[1], dest);
+}
+
+int main(int argc, char **argv)
+{
+    std::cout << argv[0] << std::endl;
+    testing::InitGoogleTest(&argc, argv);
+    return RUN_ALL_TESTS();
+}
```
Comment (garethellis0) [inline src/thunderbots/software/test/ai/navigator/path_planner/straight_line_path_planner.cpp]: github complaining about end-of-file

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Looks pretty good to me, few small comments

File: src/thunderbots/software/ai/navigator/path_planner/path_planner.h (hunk 2 above)
Comment (jordanzliu) [reply src/thunderbots/software/ai/navigator/path_planner/path_planner.h]: fixed

File: src/thunderbots/software/ai/navigator/path_planner/straight_line_path_planner.cpp (hunk 3 above)
Comment (jordanzliu) [reply src/thunderbots/software/ai/navigator/path_planner/straight_line_path_planner.cpp]: there's a start of file comment in the `straight_line_path_planner.h`

File: src/thunderbots/software/ai/navigator/path_planner/straight_line_path_planner.h (hunk 4 above)
Comment (jordanzliu) [reply src/thunderbots/software/ai/navigator/path_planner/straight_line_path_planner.h]: fixed

File: src/thunderbots/software/ai/navigator/path_planner/violation_function.h (hunk 5 above)
Comment (jordanzliu) [reply src/thunderbots/software/ai/navigator/path_planner/violation_function.h]: fixed

File: src/thunderbots/software/ai/navigator/path_planner/violation_function.h (hunk 6 above)
Comment (jordanzliu) [reply src/thunderbots/software/ai/navigator/path_planner/violation_function.h]: made the change

File: src/thunderbots/software/test/ai/navigator/path_planner/straight_line_path_planner.cpp (hunk 7 above)
Comment (jordanzliu) [reply src/thunderbots/software/test/ai/navigator/path_planner/straight_line_path_planner.cpp]: fixed

File: src/thunderbots/software/ai/navigator/path_planner/straight_line_path_planner.cpp (hunk 3 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/navigator/path_planner/straight_line_path_planner.cpp]: Technically we're supposed to put it in the `.cpp` and the `.h`. I agree that it's probably reasonable to just start putting it in the `.h` file, but we should at least be consistent for now.

**Comment by garethellis0** [comment]
Just one comment left, It's not a huge deal, but we should try to follow through on the code review checklist (I'm totally fine if we want to change it, but we should be consistent either way).

File: src/thunderbots/software/ai/navigator/path_planner/straight_line_path_planner.cpp (hunk 3 above)
Comment (jordanzliu) [reply src/thunderbots/software/ai/navigator/path_planner/straight_line_path_planner.cpp]: added a comment

**Review by garethellis0 — APPROVED** [review]

**Comment by garethellis0** [comment]
LGTM :+1: :tada:

**Pull request merged**